File: jftt/lab3/srcs/parseState.cpp
#include "parseState.hpp"

#include <iostream>

#include "gf.hpp"

namespace parseState {

using GF = gf::GF<1234577>;
using GF2 = gf::GF<1234576>;

ParseState::ParseState() {}

ParseState::ParseState(std::string text) {
  this->value = std::stol(text);
  this->text = std::string{};
  this->isError = false;
}

ParseState ParseState::setNeg() {
  if (!this->isError) {
    this->value = GF(-this->value);
    this->text = std::to_string(this->value) + ' ';
  }

  return *this;
}

ParseState ParseState::setNum() {
  if (!this->isError) {
    this->value = GF(this->value);
    this->text = std::to_string(this->value) + ' ';
  }

  return *this;
}

ParseState ParseState::add(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF(this->value) + GF(other.value);
    this->text = this->text + other.text + "+ ";
  }

  return *this;
}

ParseState ParseState::sub(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF(this->value) - GF(other.value);
    this->text = this->text + other.text + "- ";
  }

  return *this;
}

ParseState ParseState::mul(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF(this->value) * GF(other.value);
    this->text = this->text + other.text + "* ";
  }

  return *this;
}

ParseState ParseState::div(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    try {
      this->value = GF(this->value) / GF(other.value);
      this->text = this->text + other.text + "/ ";
    } catch (gf::GFNoInverseException) {
      this->text = std::to_string(GF(other.value)) +
                   " nie jest odwracalne modulo " +
                   std::to_string(GF::MAX_VALUE);
      this->isError = true;
    }
  }

  return *this;
}

ParseState ParseState::neg() {
  if (!this->isError) {
    this->value = GF(-this->value);
  }

  return *this;
}

ParseState ParseState::exp(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF(this->value) ^ GF(other.value);
    this->text = this->text + other.text + "^ ";
  }

  return *this;
}

ParseState ParseState::setNegExp() {
  if (!this->isError) {
    this->value = GF2(-this->value);
    this->text = std::to_string(this->value) + ' ';
  }

  return *this;
}

ParseState ParseState::setNumExp() {
  if (!this->isError) {
    this->value = GF2(this->value);
    this->text = std::to_string(this->value) + ' ';
  }

  return *this;
}

ParseState ParseState::addExp(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF2(this->value) + GF2(other.value);
    this->text = this->text + other.text + "+ ";
  }

  return *this;
}

ParseState ParseState::subExp(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF2(this->value) - GF2(other.value);
    this->text = this->text + other.text + "- ";
  }

  return *this;
}

ParseState ParseState::divExp(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    try {
      this->value = GF2(this->value) / GF2(other.value);
      this->text = this->text + other.text + "/ ";
    } catch (gf::GFNoInverseException) {
      this->text = std::to_string(GF2(other.value)) +
                   " nie jest odwracalne modulo " +
                   std::to_string(GF2::MAX_VALUE);
      this->isError = true;
    }
  }

  return *this;
}

ParseState ParseState::mulExp(const ParseState& other) {
  if (!this->isError && other.isError) {
    this->text = other.text;
    this->isError = true;
  }

  if (!this->isError) {
    this->value = GF2(this->value) * GF2(other.value);
    this->text = this->text + other.text + "* ";
  }

  return *this;
}

ParseState ParseState::negExp() {
  if (!this->isError) {
    this->value = GF2(-this->value);
  }

  return *this;
}

std::ostream& operator<<(std::ostream& os, const ParseState& ps) {
  if (ps.isError) {
    os << "> " << ps.text << '\n';
  } else {
    os << ps.text << "\n= " << ps.value << '\n';
  }

  return os;
}

}  // namespace parseState

File: jftt/lab3/python/parseState.py
from gf import GF, GFNoInverseException, GF_MAX_VALUE
from gf2 import GF2, GF2NoInverseException, GF2_MAX_VALUE


class ParseState:
    value: int
    text: str
    isError: bool

    def __init__(self, text: str):
        self.value = int(text)
        self.text = ""
        self.isError = False

    def setNeg(self) -> "ParseState":
        if not self.isError:
            self.value = GF(-self.value).value()
            self.text = str(self.value) + " "

        return self

    def setNum(self) -> "ParseState":
        if not self.isError:
            self.value = GF(self.value).value()
            self.text = str(self.value) + " "

        return self

    def add(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF(self.value) + GF(other.value)).value()
            self.text += other.text + "+ "

        return self

    def sub(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF(self.value) - GF(other.value)).value()
            self.text += other.text + "- "

        return self

    def mul(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF(self.value) * GF(other.value)).value()
            self.text += other.text + "* "

        return self

    def div(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            try:
                self.value = (GF(self.value) / GF(other.value)).value()
                self.text += other.text + "/ "
            except GFNoInverseException:
                self.text = (
                    str(GF(other.value).value())
                    + " nie jest odwracalne modulo "
                    + str(GF_MAX_VALUE)
                )
                self.isError = True

        return self

    def neg(self) -> "ParseState":
        if not self.isError:
            self.value = GF(-self.value).value()
            self.text += "~ "

        return self

    def exp(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF(self.value) ** GF(other.value)).value()
            self.text += other.text + "^ "

        return self

    def setNegExp(self) -> "ParseState":
        if not self.isError:
            self.value = GF2(-self.value).value()
            self.text = str(self.value) + " "

        return self

    def setNumExp(self) -> "ParseState":
        if not self.isError:
            self.value = GF2(self.value).value()
            self.text = str(self.value) + " "

        return self

    def addExp(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF2(self.value) + GF2(other.value)).value()
            self.text += other.text + "+ "

        return self

    def subExp(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF2(self.value) - GF2(other.value)).value()
            self.text += other.text + "- "

        return self

    def mulExp(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            self.value = (GF2(self.value) * GF2(other.value)).value()
            self.text += other.text + "* "

        return self

    def divExp(self, other: "ParseState") -> "ParseState":
        if (not self.isError) and other.isError:
            self.text = other.text
            self.isError = True

        if not self.isError:
            try:
                self.value = (GF2(self.value) / GF2(other.value)).value()
                self.text += other.text + "/ "
            except GF2NoInverseException:
                self.text = (
                    str(GF2(other.value).value())
                    + " nie jest odwracalne modulo "
                    + str(GF2_MAX_VALUE)
                )
                self.isError = True

        return self

    def negExp(self) -> "ParseState":
        if not self.isError:
            self.value = GF2(-self.value).value()
            self.text += "~ "

        return self

    def __str__(self):
        if self.isError:
            return "> {}".format(self.text)
        else:
            return "{}\n= {}".format(self.text, self.value)
